### PR TITLE
Adds Monitoring v3 Samples

### PR DESCRIPTION
**Pull request by waprin**
This is a migration from GoB, you already reviewed it.

Note: I know this needs tests. Create an issue and assign it to me. I had punted it but I will get it done ASAP after launch.

### CODE REVIEW

**Comment by broady** [comment]
Should this be in a subdirectory? It should be, if there are ever going to be any other monitoring samples.

Please use this license header:

```
// Copyright 2015 Google Inc. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file.
```

File: monitoring/custom_metric.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+// limitations under the License.
+
+// This example shows how to create a custom metric and write TimeSeries value
+// to it. It writes a GAUGE measurement, which is a measure of value at a
+// specific point in time. This means the startTime and endTime of the interval
+// are the same. To make it easier to see the output, a random value is written.
+// When reading the TimeSeries back, a window of the last 5 minutes is used.
+// See README.md for instructions on how to run.
+
+package main
```
Comment (broady) [inline monitoring/custom_metric.go]: Remove newline above "package"

Package-level comment should be something like:

// Command <program name> ...

File: monitoring/custom_metric.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+	"time"
+
+	"math/rand"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/monitoring/v3"
+)
+
+// createCustomMetric creates a custom metric specified by metricName
```
Comment (broady) [inline monitoring/custom_metric.go]: Full-stops at the end of each comment.

File: monitoring/custom_metric.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+		TimeSeries: []*monitoring.TimeSeries{&timeseries}}
+
+	log.Printf("writeTimeseriesRequest: %s\n", formatResource(createTimeseriesRequest))
+	_, err := s.Projects.TimeSeries.
+		Create(projectResource, &createTimeseriesRequest).Do()
+	if err != nil {
+		return fmt.Errorf("Could not write time series value, %v ", err)
+	} else {
+		return nil
+	}
```
Comment (broady) [inline monitoring/custom_metric.go]: Remove `else` block.

File: monitoring/custom_metric.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+	for resp == nil || resp.MetricDescriptors == nil {
+		resp, err = getCustomMetric(svc, projectResource, metricType, metricName)
+		if err != nil {
+			log.Fatal(err)
+		}
+		time.Sleep(2 * time.Second)
+	}
+	rand.Seed(time.Now().UTC().UnixNano())
+	// write a TimeSeries value for that metric
+	err = writeTimeSeriesValue(svc, projectResource, metricType, metricName)
```
Comment (broady) [inline monitoring/custom_metric.go]:
```
if err := ...; err != nil {
...
}
```

File: monitoring/custom_metric.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+	if err != nil {
+		log.Fatal(err)
+	}
+	time.Sleep(2 * time.Second)
```
Comment (broady) [inline monitoring/custom_metric.go]: Why sleep here?

File: monitoring/list_resources.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+func main() {
+	// walk through the basic calls of the Monitoring API
+	client, err := google.DefaultClient(
+		oauth2.NoContext,
+		"https://www.googleapis.com/auth/cloud-platform",
+		"https://www.googleapis.com/auth/monitoring",
+		"https://www.googleapis.com/auth/monitoring.read",
+		"https://www.googleapis.com/auth/monitoring.write",
```
Comment (broady) [inline monitoring/list_resources.go]: Use these constants:
https://godoc.org/google.golang.org/api/monitoring/v3#pkg-constants

File: monitoring/list_resources.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+	)
+	if err != nil {
+		log.Fatal(err)
+	}
+	if len(os.Args) < 2 {
+		fmt.Println("Usage: auth.go <project_id>")
+		return
+	}
+	projectResource := "projects/" + os.Args[1]
+	// Create the Google Cloud Monitoring Service
```
Comment (broady) [inline monitoring/list_resources.go]: Remove this comment

File: monitoring/util.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+// When reading the TimeSeries back, a window of the last 5 minutes is used.
+// See README.md for instructions on how to run.
+
+package main
+
+import (
+	"encoding/json"
+)
+
+// printResource prints out our API response objects as JSON.
```
Comment (broady) [inline monitoring/util.go]: This does not print anything. It's just a simple wrapper over `MarshalIndent`.

File: monitoring/list_resources.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+		log.Fatal(err)
+	}
+	err = listMetricDescriptors(s, projectResource)
+	if err != nil {
+		log.Fatal(err)
+	}
+	err = listTimeSeries(s, projectResource)
+	if err != nil {
+		log.Fatal(err)
+	}
```
Comment (broady) [inline monitoring/list_resources.go]: Collapse these if statements, too.

`if err := ...; err != nil {`

File: monitoring/custom_metric.go [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+// See README.md for instructions on how to run.
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"time"
+
```
Comment (broady) [inline monitoring/custom_metric.go]: Remove newline

File: monitoring/README.md [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    * Click 'New Credentials', and create a Service Account or [click  here](https://console.developers.google.com/project/_/apiui/credential/serviceaccount)
+     Download the JSON for this service account, and set the `GOOGLE_APPLICATION_CREDENTIALS`
+     environment variable to point to the file containing the JSON credentials.
+
+    ```
+    export GOOGLE_APPLICATION_CREDENTIALS=~/Downloads/<project-id>-0123456789abcdef.json
+    ```
+
+# Set Up Your Local Dev Environment
+To install your dependencies, first make sure your GOPATH is set, then install the Google Go Api
```
Comment (broady) [inline monitoring/README.md]: `API`

File: monitoring/README.md [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+    go get -u google.golang.org/api/...
+
+To run the example that prints the environment, run:
+
+    go run util.go print_environment.go <your-project-id>
+
+To run the example that creates a custom metric and writes to it, run:
+
+    go run util.go custom_metric.go <your-project-id>
```
Comment (broady) [inline monitoring/README.md]: There are two main functions in the one package. Move these into two separate subdirectories.

`monitoring/sample1`
`monitoring/sample2`

File: monitoring/custom_metric.go (hunk 2 above)
Comment (waprin) [inline monitoring/custom_metric.go]: done

File: monitoring/custom_metric.go (hunk 10 above)
Comment (waprin) [inline monitoring/custom_metric.go]: done

File: monitoring/README.md (hunk 11 above)
Comment (waprin) [inline monitoring/README.md]: done

**Comment by broady** [comment]
merged with some changes: 9d9c1d10d4bc0e768b83d62787aa1e8bfa6db072

**Pull request closed without merging**